pkg/server/controllers.go

```diff
@@ -95,6 +115,69 @@ type controllerWrapper struct {
 	Wait   WaitFunc
 }
 
+func indexUnschedulable(obj interface{}) ([]string, error) {
+	workspace := obj.(*tenancyv1alpha1.Workspace)
+	if conditions.IsFalse(workspace, tenancyv1alpha1.WorkspaceScheduled) && conditions.GetReason(workspace, tenancyv1alpha1.WorkspaceScheduled) == tenancyv1alpha1.WorkspaceReasonUnschedulable {
+		return []string{"true"}, nil
+	}
+	return []string{}, nil
+}
+
+func indexByBase36Sha224Name(obj interface{}) ([]string, error) {
+	s := obj.(*corev1alpha1.Shard)
+	return []string{workspace.ByBase36Sha224NameValue(s.Name)}, nil
+}
+
+// indexAPIExportsByAPIResourceSchemasFunc is an index function that maps an APIExport to its spec.latestResourceSchemas.
+func indexAPIExportsByAPIResourceSchemasFunc(obj interface{}) ([]string, error) {
+	apiExport, ok := obj.(*apisv1alpha1.APIExport)
+	if !ok {
+		return []string{}, fmt.Errorf("obj is supposed to be an APIExport, but is %T", obj)
+	}
+
+	ret := make([]string, len(apiExport.Spec.LatestResourceSchemas))
+	for i := range apiExport.Spec.LatestResourceSchemas {
+		ret[i] = client.ToClusterAwareKey(logicalcluster.From(apiExport).Path(), apiExport.Spec.LatestResourceSchemas[i])
+	}
+
+	return ret, nil
+}
+
+// indexAPIExportEndpointSliceByAPIExportFunc indexes the APIExportEndpointSlice by their APIExport's Reference Path and Name.
+func indexAPIExportEndpointSliceByAPIExportFunc(obj interface{}) ([]string, error) {
+	apiExportEndpointSlice, ok := obj.(*apisv1alpha1.APIExportEndpointSlice)
+	if !ok {
+		return []string{}, fmt.Errorf("obj %T is not an APIExportEndpointSlice", obj)
+	}
+
+	path := logicalcluster.NewPath(apiExportEndpointSlice.Spec.APIExport.Path)
+	if path.Empty() {
+		path = logicalcluster.From(apiExportEndpointSlice).Path()
+	}
+	return []string{path.Join(apiExportEndpointSlice.Spec.APIExport.Name).String()}, nil
+}
+
+// indexAPIExportEndpointSlicesByPartitionFunc is an index function that maps a Partition to the key for its
+// spec.partition.
+func indexAPIExportEndpointSlicesByPartitionFunc(obj interface{}) ([]string, error) {
+	slice, ok := obj.(*apisv1alpha1.APIExportEndpointSlice)
+	if !ok {
+		return []string{}, fmt.Errorf("obj is supposed to be an APIExportEndpointSlice, but is %T", obj)
+	}
+
+	if slice.Spec.Partition != "" {
+		clusterName := logicalcluster.From(slice).Path()
+		if !ok {
+			// this will never happen due to validation
+			return []string{}, fmt.Errorf("cluster information missing")
+		}
+		key := client.ToClusterAwareKey(clusterName, slice.Spec.Partition)
+		return []string{key}, nil
+	}
+
+	return []string{}, nil
+}
+
```

Comment:
would move all of these into an index.go. This file here is getting big.

Reply:
Sure , we wil do